modules/gapi/src/backends/ie/giebackend.cpp

```diff
@@ -616,55 +774,115 @@ static void configureInputInfo(const IE::InputInfo::Ptr& ii, const cv::GMetaArg
 
 // NB: This is a callback used by async infer
 // to post outputs blobs (cv::GMat's).
-struct PostOutputs {
-    // NB: Should be const to pass into SetCompletionCallback
-    void operator()() const {
-        for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-             auto& out_mat = ctx->outMatR(i);
-             IE::Blob::Ptr this_blob = iec.this_request.GetBlob(ctx->uu.params.output_names[i]);
-             copyFromIE(this_blob, out_mat);
-             ctx->out.post(ctx->output(i));
-        }
-        ctx->sync.release_and_notify();
+static void PostOutputs(IECompiled                     &iec,
+                        std::shared_ptr<IECallContext>  ctx) {
+    for (auto i : ade::util::iota(ctx->uu.params.num_out))
+    {
+        auto& out_mat = ctx->outMatR(i);
+        IE::Blob::Ptr this_blob = iec.this_request.GetBlob(ctx->uu.params.output_names[i]);
+        copyFromIE(this_blob, out_mat);
+        ctx->out.post(ctx->output(i));
     }
+}
 
-    IECompiled                            &iec ;
-    std::shared_ptr<IECallContext>         ctx ;
-};
+// NB: This is a callback used by InferList
+class PostInferList {
+public:
+    PostInferList(std::shared_ptr<IECallContext> ctx,
+                  std::vector<std::vector<int>>  cached_dims,
+                  std::vector<cv::Rect>          in_roi_vec,
+                  IE::Blob::Ptr                  this_blob,
+                  cv::gimpl::ie::RequestPool     &reqPool);
 
-// NB: This is a callback used by async infer
-// to post output list of blobs (cv::GArray<cv::GMat>).
-struct PostOutputsList {
-    // NB: Should be const to pass into SetCompletionCallback
-    void operator()() const {
-        for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-            std::vector<cv::Mat> &out_vec = ctx->outVecR<cv::Mat>(i);
+    // NB: Should be const to use it as a callback in InferenceEngine.
+    void operator()(IECompiled& iec) const;
 
-            IE::Blob::Ptr out_blob = iec.this_request.GetBlob(ctx->uu.params.output_names[i]);
+private:
+    struct Priv;
+    explicit PostInferList(std::shared_ptr<Priv> p);
 
-            cv::Mat out_mat(cached_dims[i], toCV(out_blob->getTensorDesc().getPrecision()));
-            // FIXME: Avoid data copy. Not sure if it is possible though
-            copyFromIE(out_blob, out_mat);
-            out_vec.push_back(std::move(out_mat));
-        }
-        // NB: Callbacks run synchronously yet, so the lock isn't necessary
-        auto&& out_vec_size = ctx->outVecR<cv::Mat>(0).size();
-        // NB: Now output vector is collected and can be posted to output
-        if (nrequests == out_vec_size) {
-            for (auto i : ade::util::iota(ctx->uu.params.num_out)) {
-                 ctx->out.post(ctx->output(i));
-            }
-        }
+    std::shared_ptr<Priv> m_priv;
+};
 
-        ctx->sync.release_and_notify();
-    }
+struct PostInferList::Priv {
```

Comment:
Do you really need Priv here?

Reply:
Yes, I explained this in code.
 Briefly, `PostInferList` in case not all items ('roi'`) have been processed, put the task to process next `roi`, but the state of the callback have to be alive because it will be used in the next callback. Just extend life cycle by increase reference counter.

Reply:
Not relevant